Update padFootprint function

Nisarg236 opened the pull request:
[AB#44902](https://dev.azure.com/rapyuta-robotics/sootballs/_workitems/edit/44902)
Previously this function was scaling the footprint with respect to the coordinate frame 0,0 which could result in distortion. Also when the centroid is having large offset with 0,0 then it will just displace the footprint instead of padding.

Now it scales with respect to the centroid of the footprint.

![image](https://github.com/user-attachments/assets/3d2ad01c-c556-46da-b05c-e648a352e6e7)
![image](https://github.com/user-attachments/assets/b9c54097-5db0-4f5b-af45-c72ac897428d)
another case (10,10), (10,12), (12,12) , (12,10)
![image](https://github.com/user-attachments/assets/8f4bf020-c5ef-4687-8061-a80d297f02ac)
![image](https://github.com/user-attachments/assets/7fb65778-9d7f-4af7-8188-1fbee38888c8)

another case (0,0), (0,2), (2,2), (2,0)
![image](https://github.com/user-attachments/assets/a5340025-3ebd-4f5c-9b84-6e7440267e29)
![image](https://github.com/user-attachments/assets/9a86cdc0-1cee-471b-9e0d-03ca7b8b20d6)

I have also tested the behaviour with negative values of padding (shrinking the footprint)

Nisarg236 committed:
scale wrt centroid

Nisarg236 committed:
minor fix

Nisarg236 committed:
update unit tests

Nisarg236 committed:
update unit tests

corot reviewed (changes requested):
3 things

The 1st example looks suspicious:  doesn't look like padded 10 cm but 2 or 3  :thinking: 

This looks similar to [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d)
Do both provide the same result?  If so, this is simpler

Keep the formatting of the files you modify. i.e. never mix stiles

Nisarg236 committed:
scale by meters not percentage

Nisarg236 commented:
> 3 things
> 
> The 1st example looks suspicious: doesn't look like padded 10 cm but 2 or 3 🤔
> 
> This looks similar to [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d) Do both provide the same result? If so, this is simpler
> 
> Keep the formatting of the files you modify. i.e. never mix stiles

Hi you are right it wasnt 0.1 meters because it was scaling by percentage, I changed it to scale by meters and updated the image.
and now the functionality is the same as the one you shared.

corot commented:
> this commit

corot commented:
> Hi you are right it wasnt 0.1 meters because it was scaling by percentage, I changed it to scale by meters and updated the image. and now the functionality is the same as the one you shared.

So the outcome is now the same as with [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d)?  
Then better to use the commit,,, as the logic is much simpler  :thinking:

renan028 commented:
> > Hi you are right it wasnt 0.1 meters because it was scaling by percentage, I changed it to scale by meters and updated the image. and now the functionality is the same as the one you shared.
> 
> So the outcome is now the same as with [this commit](https://github.com/corot/navigation/commit/7c4fbb9338178a240425f9b93efa994b0bb0f00d)? Then better to use the commit,,, as the logic is much simpler 🤔

@corot I think you are missing the part he calculates the centroid of the polygon.

renan028 commented on costmap_2d/src/footprint.cpp:
```diff
@@ -137,12 +137,29 @@ void transformFootprint(double x, double y, double theta, const std::vector<geom
 {
-  // pad footprint in place
-  for (unsigned int i = 0; i < footprint.size(); i++)
+  geometry_msgs::Point footprint_center;
+  double sum_x = 0.0, sum_y = 0.0;
+
+  for (const auto& point : footprint)
+  {
+    sum_x += point.x;
+    sum_y += point.y;
+  }
+  footprint_center.x = sum_x / footprint.size();
+  footprint_center.y = sum_y / footprint.size();
```
S: I think it makes sense to create a method `centroid` that returns `geometry_msgs::Point`

renan028 commented on costmap_2d/src/footprint.cpp:
```diff
@@ -137,12 +137,29 @@ void transformFootprint(double x, double y, double theta, const std::vector<geom
+    double dx = point.x - footprint_center.x;
+    double dy = point.y - footprint_center.y;
+
+    double distance = std::sqrt(dx * dx + dy * dy);
```
S: `std::hypot`

renan028 commented:
I think it makes sense to make a PR upstream:
https://github.com/ros-navigation/navigation2/blob/4e5d2dfc66cb75eb390d614d13e5b64efbf30284/nav2_costmap_2d/src/footprint.cpp#L145

corot commented on costmap_2d/test/footprint_tests.cpp:
```diff
@@ -70,17 +70,17 @@ TEST( Costmap2DROS, padded_footprint_from_string_param )
-  EXPECT_EQ( 1.5f, footprint[ 0 ].x );
-  EXPECT_EQ( 1.5f, footprint[ 0 ].y );
-  EXPECT_EQ( 0.0f, footprint[ 0 ].z );
+  EXPECT_EQ( 1.66667f, footprint[0].x );
```
Q: do we really need to change round numbers?

Nisarg236 replied:
yes because otherwise it fails

corot reviewed (approved):
ok, as @renan028 explained, this implementation is superior to the one I mention

@Nisarg236 , you should consider open a PR to nav2, as this is a nice contribution

Nisarg236 committed:
separate function for centroid and use std::hypot

Nisarg236 committed:
update test case

Nisarg236 committed:
update test case
